DEVELOPMENT.md

```diff
@@ -0,0 +1,229 @@
+
+- Mocks: `make mock-gen`
+- Protobuf: `make proto-gen`
+- Thrift: `make thrift-gen`
+
+Don't forget to account for changes to generated files in tests!
+
+## Testing Changes
+
+M3 has an extensive test suite to ensure that we are able to validate changes.
+More notes about the various testing strategies we employ can be found in
+[TESTING][TESTING.md].
+
+While our CI suite will run all tests prior to allowing pull requests to be
+merged, developers should test their changes during development and prior to
+pushing updates. To test code, simply use `go test`:
+
+```
+go test ./... -v
+```
+
+Unfortunately, as more tests have been added, the time required to run the
+entire suite has increased significantly; therefore, we recommend at least
+running tests for both (1) any packages with changes and (2) any packages that
+import the changed packages. A crude example of this would be:
+
+```bash
+changed=$(git diff --name-only HEAD^ HEAD | xargs -I {} dirname {} | sort | uniq)
+for pkg in $changed; do
+  affected=$(grep -r "$pkg" ./src | cut -d: -f1 | grep -v mock | xargs -I{} dirname {} | sort | uniq)
+  go test -v -race "./$pkg" $affected
+done
+```
+
+Of course, contributors are free to do whatever due diligence they feel helps
+them to be most productive. Our only request is that CI jobs are not used as a
+first-pass filter to determine whether a change is sane or not.
+
+Once tests are passing locally, push to a new remote branch to create a pull
+request, or push to an existing branch to update a pull request. If the CI
+suite reports any errors, attempt to reproduce failures locally and fix them
+before continuing.
+
+For larger or more intensive tests (e.g. "big" unit tests, integration tests),
+additional build tags may be necessary to scope the tests down to a smaller
+subset, e.g.:
+
+```
+# example integration test
+$ go test ./integration -tags integration -run TestIndexBlockRotation -v
+
+# example big unit test
+$ go test -tags big ./services/m3dbnode/main -run TestIndexEnabledServer -v
+```
+
+## Code Review
+
+Please observe the following guidelines when submitting or reviewing pull
+requests:
+
+- Merging is blocked on approval by 1 or more users with write access. We
+  recommend 2+ for large, complex, or nuanced changes.
+- Pull requests should contain clear descriptions and context per the
+  pull request template.
+- Breaking changes under the current versioning guarantees (see
+  [COMPATIBILITY][COMPATIBILITY.md]) must be approved by the
+  [TSC][GOVERNANCE.md].
+- The [STYLEGUIDE][STYLEGUIDE.md] should be followed to the extent reasonable
+  within the scope of the pull request.
+- Changed codepaths should be exercised by unit tests that cover at least one
+  nominal case and one error case.
+- Pull requests may only be merged with a green build. Do not merge with
+  build failures, even if they are known and unrelated.
+- Flaky or otherwise spurious CI failures count as hard failures. Commit fixes
```

Comment:
```suggestion
- Flaky or otherwise unreliable CI failures count as hard failures. Commit fixes
```

Spurious is not the best word for non-native readers